### PR TITLE
feat(sectionedForm): initial architecture and base components

### PR DESCRIPTION
**Pull request by Birkbjo**
https://dhis2.atlassian.net/browse/DHIS2-18334

Implements base components for `SectionedForm`s. These are forms with a sidebar, with links to different sections.
In a recent meeting it was decided we probably want to keep most section in a "scrollable"-view - ie. all fields should render. 

This also implements some fields for `dataSet` "New"-forms, to showcase the use of the components. 

- Introduces a concept of `formDescriptor`. This is put in to context to allow sub-component to retrieve it.
   - The main purpose of this is to able to get labels for fields and sections, as well as describe which fields belong to what section. 
   - I believe it's important to have the ability to imperatively be able to fetch which section a certain field belongs to. This is useful for showing error-indicators in the "Section-sidebar" to communicate to the user which section has errors. It's also useful to show a summary of errors fields - because previously the label was grabbed through the DOM, which is not really ideal.
   - The context also includes some helper-functions, to eg. get a field label. The structure needed to keep order of sections/fields, so regular keyed objects couldn't be used to key them.

### CODE REVIEW

File: src/lib/form/sectionedForm/useSelectedSection.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { useMemo } from 'react'
+import { useQueryParam, createEnumParam, withDefault } from 'use-query-params'
+import { useSectionedFormDescriptor } from './SectionedFormContext'
+
+export const FORM_SECTION_PARAM_KEY = 'section'
+
+export const getSectionSearchParam = (section: string) => {
```
Comment (Birkbjo) [inline src/lib/form/sectionedForm/useSelectedSection.ts:7]: Unused

File: src/components/sectionedForm/SectionedFormFooter.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    return <div className={css.formActions}>{children}</div>
+}
+
+const SectionedFormFooterSectionActions = ({
+    children,
+}: React.PropsWithChildren) => {
+    return <div className={css.sectionActions}>{children}</div>
+}
+
+const VerticalDivider = () => <span className={css.verticalDivider}></span>
```
Comment (Birkbjo) [inline src/components/sectionedForm/SectionedFormFooter.tsx:16]: Unused

File: src/components/form/FormBase.tsx [hunk 3]
```diff
@@ -15,14 +19,17 @@ type MaybeModelWithAttributes = {
 
 type OwnProps<TValues = Record<string, unknown>> = {
     initialValues: TValues | undefined
-    children: React.ReactNode
+    children: FormProps<TValues>['children']
```
Comment (Birkbjo) [inline src/components/form/FormBase.tsx:22]: May need an union here

File: src/components/NavigationBlocker/DefaultFormNavigationBlocker.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    ModalContent,
+    ModalTitle,
+    ButtonStrip,
+    Button,
+} from '@dhis2/ui'
+import React from 'react'
+import { useFormState } from 'react-final-form'
+import { useBlocker } from 'react-router-dom'
+
+export const DefaultFormNavigationBlocker = () => {
```
Comment (Birkbjo) [inline src/components/NavigationBlocker/DefaultFormNavigationBlocker.tsx:14]: Wasn't my intention to add this actually - it's not used in any form yet. So I can remove from this PR and add later if we want.

File: src/components/form/DefaultFormErrorNotice.tsx [hunk 5]
```diff
@@ -1,57 +1,51 @@
...
-    const partialFormState: FormStateErrors = useFormState({
-        subscription: formStateSubscriptions,
-    })
-    // only show after trying to submit
-    if (
-        !partialFormState.submitFailed ||
-        (partialFormState.submitFailed && partialFormState.dirtySinceLastSubmit)
-    ) {
+    const formStateErrors = useFormStateErrors()
+    console.log(formStateErrors)
```
Comment (Birkbjo) [inline src/components/form/DefaultFormErrorNotice.tsx:9]: remove

File: src/components/sectionedForm/SectionedFormErrorNotice.tsx [hunk 6]
```diff
@@ -30,16 +37,22 @@ export function SectionedFormErrors() {
...
+                <div className={css.errorNoticeBoxHeader}>
+                    <h6>{i18n.t('Validation errors')}</h6>
+                    <span onClick={() => setClosed((prev) => !prev)}>
+                        <IconCross16 />
+                    </span>
                 </div>
-                <div>
+                <div className={css.errorNoticeBoxContent}>
+                    <ErrorList errors={errorsWithlabels} />
+                    <ErrorList errors={errorsWithlabels} />
```
Comment (Birkbjo) [inline src/components/sectionedForm/SectionedFormErrorNotice.tsx:50]: Cleanup testing code

File: src/lib/form/sectionedForm/useSelectedSection.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        [originalHandler]
+    )
+    return [selected, setHandler] as const
+}
+
+/**
+ * Update the selected section (in searchParams) based on the section that is in view
+ * This keeps the selected section in sync with the section that is in view.
+ */
+export const useSyncSelectedSectionWithScroll = () => {
```
Comment (Birkbjo) [inline src/lib/form/sectionedForm/useSelectedSection.ts:81]: Not super happy about this, but I couldn't find a better way.

**Review by flaminic — APPROVED** [review]

**Pull request merged**